utils.py

```diff
@@ -83,44 +88,102 @@ def MakePlots(truth_XY, predict_XY, PUPPI_XY, path_out):
+    bin_resolX_PUPPI, _, _ = binned_statistic(true_ptPhi[:,0], trueXY[:,0] - puppiXY[:,0] * sfs_PUPPI, statistic=resolqt, bins=binnings, range=(0,400))
+    bin_resolY_PUPPI, _, _ = binned_statistic(true_ptPhi[:,0], trueXY[:,1] - puppiXY[:,1] * sfs_PUPPI, statistic=resolqt, bins=binnings, range=(0,400))
+
+    # and the upper and lower bounds
+    bin_resolX_ML_upper, bin_edges, binnumber = binned_statistic(true_ptPhi[:,0], trueXY[:,0] - mlXY[:,0] * sfs_ML, statistic=resolqt_upper, bins=binnings, range=(0,400))
+    bin_resolY_ML_upper, _, _ = binned_statistic(true_ptPhi[:,0], trueXY[:,1] - mlXY[:,1] * sfs_ML, statistic=resolqt_upper, bins=binnings, range=(0,400))
+    bin_resolX_PUPPI_upper, _, _ = binned_statistic(true_ptPhi[:,0], trueXY[:,0] - puppiXY[:,0] * sfs_PUPPI, statistic=resolqt_upper, bins=binnings, range=(0,400))
+    bin_resolY_PUPPI_upper, _, _ = binned_statistic(true_ptPhi[:,0], trueXY[:,1] - puppiXY[:,1] * sfs_PUPPI, statistic=resolqt_upper, bins=binnings, range=(0,400))
+
+    bin_resolX_ML_lower, bin_edges, binnumber = binned_statistic(true_ptPhi[:,0], trueXY[:,0] - mlXY[:,0] * sfs_ML, statistic=resolqt_lower, bins=binnings, range=(0,400))
+    bin_resolY_ML_lower, _, _ = binned_statistic(true_ptPhi[:,0], trueXY[:,1] - mlXY[:,1] * sfs_ML, statistic=resolqt_lower, bins=binnings, range=(0,400))
+    bin_resolX_PUPPI_lower, _, _ = binned_statistic(true_ptPhi[:,0], trueXY[:,0] - puppiXY[:,0] * sfs_PUPPI, statistic=resolqt_lower, bins=binnings, range=(0,400))
+    bin_resolY_PUPPI_lower, _, _ = binned_statistic(true_ptPhi[:,0], trueXY[:,1] - puppiXY[:,1] * sfs_PUPPI, statistic=resolqt_lower, bins=binnings, range=(0,400))
+
+    # calculate ML and PUPPI resolutions weighted difference
+    weights = []
+    nEvents = len(binnumber)
+    for bin in np.arange(len(binnings)-1):
+        weights.append(len(binnumber[binnumber==bin+1])/nEvents) #number of events in bin / total number of events
+    xRes_avgDif = np.average(bin_resolX_PUPPI-bin_resolX_ML, weights=weights)
+    yRes_avgDif = np.average(bin_resolY_PUPPI-bin_resolY_ML, weights=weights)
+
+    #make the plots
     plt.figure()
-    plt.hlines(bin_resolX_predict, bin_edges[:-1], bin_edges[1:], colors='r', lw=5,
-           label='Predict', linestyles='solid')
+    plt.hlines(bin_resolX_ML, bin_edges[:-1], bin_edges[1:], colors='r', lw=5,
+               label='ML', linestyles='solid')
     plt.hlines(bin_resolX_PUPPI, bin_edges[:-1], bin_edges[1:], colors='g', lw=5,
-           label='PUPPI', linestyles='solid')
+               label='PUPPI', linestyles='solid')
+    xRes_ML_error = abs([bin_resolX_ML_lower, bin_resolX_ML_upper]-bin_resolX_ML)
+    plt.errorbar(bin_edges[:-1]+12, bin_resolX_ML, yerr= xRes_ML_error, fmt='none', color='r')
+    xRes_PUPPI_error = abs([bin_resolX_PUPPI_lower, bin_resolX_PUPPI_upper]-bin_resolX_PUPPI)
+    plt.errorbar(bin_edges[:-1]+8, bin_resolX_PUPPI, yerr= xRes_PUPPI_error, fmt='none', color='g')
```

Comment:
8/12 are magic numbers

Reply:
+10 will center the error bars on the horizontal lines. +8 +12 puts them close without overlapping

Reply:
could you add comment or better yet use variable names to define this:

`binWidth = 10`, etc.